File: day-2/part-1/elixir/coordinates.exs
defmodule Coordinates do

  def calculate(input) do
    directions = prepare_directions(input)
    directions
      |> Enum.reduce({0, 0}, fn direction, acc ->
        move(direction, acc)
      end)
      |> Tuple.product()
  end

  defp prepare_directions(input) when is_binary(input) do
    input
      |> String.split("\n")
      |> Enum.map(&String.trim/1)
      |> Enum.filter(&(String.length(&1) > 0))
  end

  defp prepare_directions(file) do
    case file do
      {:ok, content} -> prepare_directions(content)
      {:error, _} -> []
    end
  end

  defp move(<<"forward ", number::binary>>, {x, y}) do
    {number, _} = Integer.parse(number)
    {x + number, y}
  end

  defp move(<<"up ", number::binary>>, {x, y}) do
    {number, _} = Integer.parse(number)
    {x, y - number}
  end

  defp move(<<"down ", number::binary>>, {x, y}) do
    {number, _} = Integer.parse(number)
    {x, y + number}
  end
end

File: day-2/part-1/ruby/coordinates.rb
class Coordinates
  def calculate(input)
    directions = prepare_directions(input)
    directions.reduce({x: 0, y: 0}) do |acc, direction|
      move(direction, acc)
    end.values.inject(1, &:*)
  end

  private

  def move(direction, position)
    position => {x:, y:}
    cor, number = direction.split(' ')
    case cor
    when 'forward'
      {x: x + number.to_i, y: y}
    when 'up'
      {x: x, y: y - number.to_i}
    when 'down'
      {x: x, y: y + number.to_i}
    end
  end

  def prepare_directions(file)
    file.split("\n")
  end
end